### PR TITLE
WIP: Add JIRA backend

### PR DESCRIPTION
**Pull request by nkprince007**

### CODE REVIEW

**Comment by nkprince007** [comment]
@omab, some help here, please?

File: requirements-base.txt [hunk 1]
```diff
@@ -3,3 +3,4 @@ oauthlib>=1.0.3
 requests-oauthlib>=0.6.1
 six>=1.10.0
 PyJWT>=1.4.0
+cryptography>=2.1.1
```
Comment (omab) [inline requirements-base.txt]: Where's this used?

File: requirements-python3.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
 defusedxml>=0.5.0rc1
 python3-openid>=3.0.10
+pycryptodome>=3.4.7
```
Comment (omab) [inline requirements-python3.txt]: Where's this used?

File: social_core/backends/jira.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    def get_unauthorized_token(self):
+        return self.get_json(
+            self.REQUEST_TOKEN_URL,
+            auth=self.oauth_auth()
+        )
+
+    @property
+    def ACCESS_TOKEN_URL(self):
+        return self._ACCESS_TOKEN_URL.format(**self.jira_url_params)
```
Comment (omab) [inline social_core/backends/jira.py]: There's no need to define this property, you can define the method `access_token_url` instead.

File: social_core/backends/jira.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    @property
+    def AUTHORIZATION_URL(self):
+        return self._AUTHORIZATION_URL.format(**self.jira_url_params)
```
Comment (omab) [inline social_core/backends/jira.py]: There's no need to define this property, you can define the method `authorization_url` instead.

File: requirements-python3.txt (hunk 2 above)
Comment (nkprince007) [reply requirements-python3.txt]: It's being used for RSA1 encryption.

Comment (nkprince007) [reply requirements-python3.txt]: It is an external dependency which oauthlib isn't bundled with.

File: requirements-base.txt (hunk 1 above)
Comment (nkprince007) [reply requirements-base.txt]: Please check https://github.com/python-social-auth/social-core/pull/155#discussion_r157049337

**Comment by nkprince007** [comment]
@omab please check the replies on review comments. I've updated the pull request with the changes you've suggested.

Now for handling JIRA Server REST APIs, each request sent to the server should be signed by an RSA priv/pub key pair which has already been registered as an incoming Application link in JIRA administration.

So, these additional requirements, `pycryptodome` and `cryptography` are required for RSA 1 signing. These algorithms are not bundled along with `oauthlib` and need to be installed separately.

**Pull request closed without merging**